nasa-gibs/worldview
# WV-2223: Only show granules that have imagery

## Description

These changes allow granule layers to have a different start date, end date, and date ranges, as defined in the GetCapabilities, depending on the projection.  Currently we are relying on the string `Granule` to appear in the layer id in order to determine whether or not to format the layer config output in this way.  If a layer has different dates in different projections the output will look like:

```
'VIIRS_NOAA20_CorrectedReflectance_TrueColor_Granule_v1_NRT': {
        id: 'VIIRS_NOAA20_CorrectedReflectance_TrueColor_Granule_v1_NRT',
        group: 'overlays',
        projections: {
          geographic: {
            startDate: '2019-09-23T00:12:00Z',
            endDate: '2019-09-24T23:54:00Z',
            dateRanges: [... ],
          },
          arctic: {
            startDate: '2019-07-21T00:36:00Z',
            endDate: '2019-09-24T22:30:00Z',
            dateRanges: [...],
          },
          antarctic: {
            startDate: '2019-07-12T01:18:00Z',
            endDate: '2019-09-24T23:36:00Z',
            dateRanges: [...],
          },
        },
      },
```

Notice there are no start, end, or date range properties at the top level but instead they are nested under `projections`.  This allows us to accurately determine whether or not a granule footprint, pulled from CMR, has associated GIBS imagery in a given projection.  This also allows us to show the appropriate dates in the layer info views.

Other minor changes made:

* Consolidate dispatching actions related to projection change (dispatch side effect actions from projection change action)
* Remove unused resetLayers action

## How To Test

### Setup
Layer configs for testing are located on a separate branch

1. Checkout `wv-2223-config branch
2. `npm run build`
3. Checkout `wv-2223`
4. `npm ci`
5. `npm run start`

### Tests

1. [Load a URL with a granule layer added](http://localhost:3000/?v=-130.62103600054846,-135.74579787838871,223.0673133405011,167.93520129875924&z=4&ics=true&ici=5&icd=10&l=VIIRS_NOAA20_CorrectedReflectance_TrueColor_Granule_v1_NRT,Coastlines_15m&lg=true&t=2019-09-24-T16%3A51%3A34Z)
2. Hover over granules
3. Confirm that footprint outlines only show for granules where imagery is visible
4. Open layer info from sidebar
5. Confirm temporal coverage and date ranges match what is found in the [GC for this projection](https://uat.gibs.earthdata.nasa.gov/wmts/epsg4326/nrt/wmts.cgi?request=GetCapabilities) 
6. Switch to arctic projection
7. Hover over granules
8. Confirm that footprint outlines only show for granules where imagery is visible
9. Open layer info from sidebar
10. Confirm temporal coverage and date ranges match what is found in the [GC for this projection](https://uat.gibs.earthdata.nasa.gov/wmts/epsg3031/nrt/wmts.cgi?request=GetCapabilities)
